File: linear_search/java/LinearSearchPerf.java
import java.util.Arrays;

public class LinearSearchPerf {

    static final int N = 1_000_000_000;

    public static int[] generateArray(int size) {
        int[] arr = new int[size];
        for (int i = 0; i < size; i++) {
            arr[i] = i;
        }
        return arr;
    }

    public static int linearSearch(int[] arr, int target) {
        for (int i = 0; i < arr.length; i++) {
            if (arr[i] == target) {
                return i;
            }
        }
        return -1;
    }

    public static void benchmarkSearch(int[] arr, int index, String label) {
        int target = arr[index];
        long start = System.nanoTime();
        int result = linearSearch(arr, target);
        long end = System.nanoTime();
        double elapsedSeconds = (end - start) / 1_000_000_000.0;
        System.out.printf("%s: %.8f seconds%n", label, elapsedSeconds);
    }

    public static void main(String[] args) {
        System.out.printf("Generating array of %d integers... and searching for values located at%n", N);
        System.out.println("Technology: Java");

        int[] data = generateArray(N);

        benchmarkSearch(data, N / 100, "0%");
        benchmarkSearch(data, N / 20, "5%");
        benchmarkSearch(data, (int)((long)N * 15 / 100), "15%");
        benchmarkSearch(data, N / 4, "25%");
        benchmarkSearch(data, N / 2, "50%");
        benchmarkSearch(data, (int)((long)N * 3 / 4), "75%");
        benchmarkSearch(data, N - 1, "100%");
    }
}

File: linear_search/c/linear_search_perf.c
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <time.h>

#define N 1000000000UL

int* generate_array(size_t size) {
    int* arr = calloc(size, sizeof(int));
    if (!arr) {
        fprintf(stderr, "Memory allocation failed\n");
        exit(EXIT_FAILURE);
    }
    for (size_t i = 0; i < size; ++i)
        arr[i] = (int)i;
    return arr;
}

int linear_search(int* arr, size_t size, int target) {
    for (size_t i = 0; i < size; ++i)
        if (arr[i] == target)
            return (int)i;
    return -1;
}

void benchmark_search(int* arr, size_t size, size_t index, const char* label) {
    int target = arr[index];
    clock_t start = clock();
    linear_search(arr, size, target);
    clock_t end = clock();
    double elapsed = (double)(end - start) / CLOCKS_PER_SEC;
    printf("%s: %.8f seconds\n", label, elapsed);
}

int main() {
    printf("Generating array of %lu integers... and searching for values located at\n", (unsigned long)N);
    printf("Technology: ANSI C99\n");

    int* data = generate_array(N);

    benchmark_search(data, N, N / 100, "0%");         // 0%
    benchmark_search(data, N, N / 20, "5%");          // 5%
    benchmark_search(data, N, (size_t)N * 15 / 100, "15%"); // 15%
    benchmark_search(data, N, N / 4, "25%");          // 25%
    benchmark_search(data, N, N / 2, "50%");          // 50%
    benchmark_search(data, N, (size_t)N * 3 / 4, "75%");    // 75%
    benchmark_search(data, N, N - 1, "100%");         // 100%

    free(data);
    return 0;
}

